src/arithmetic/algebraic_expression/algebraic_expression_eval.c

```diff
@@ -0,0 +1,255 @@
+
+	uint child_count = AlgebraicExpression_ChildCount(exp);
+	// Expression has more than 2 operands, e.g. A+B+C...
+	for(uint i = 2; i < child_count; i++) {
+		right = exp->operation.children[i];
+
+		if(right->type == AL_OPERAND) {
+			b = right->operand.matrix;
+		} else {
+			if(right->operation.op == AL_EXP_TRANSPOSE) {
+				b = right->operation.children[0]->operand.matrix;
+				GrB_Descriptor_set(desc, GrB_INP1, GrB_TRAN);
+			} else if(inter == GrB_NULL) {
+				// Can't use `res`, use an intermidate matrix.
+				GrB_Matrix_nrows(&nrows, res);
+				GrB_Matrix_ncols(&ncols, res);
+				GrB_Matrix_new(&inter, GrB_BOOL, nrows, ncols);
+			}
+			b = _AlgebraicExpression_Eval(right, inter);
+		}
+
+		// Perform addition.
+		if(GrB_eWiseAdd_Matrix_Semiring(res, GrB_NULL, GrB_NULL, GxB_LOR_LAND_BOOL, res, b,
+										GrB_NULL) != GrB_SUCCESS) {
+			printf("Failed adding operands, error:%s\n", GrB_error());
+			assert(false);
+		}
+
+		// Reset descriptor.
+		GrB_Descriptor_set(desc, GrB_INP1, GxB_DEFAULT);
+	}
+
+	if(inter != GrB_NULL) GrB_Matrix_free(&inter);
+	GrB_free(&desc);
+	return res;
+}
+
+static GrB_Matrix _Eval_Mul(const AlgebraicExpression *exp, GrB_Matrix res) {
+	assert(exp &&
+		   AlgebraicExpression_ChildCount(exp) > 1 &&
+		   AlgebraicExpression_OperationCount(exp, AL_EXP_MUL) == 1);
+
+	GrB_Matrix A;
+	GrB_Matrix B;
+	GrB_Info info;
+	GrB_Index nvals;
+	GrB_Descriptor desc;
+	AlgebraicExpression *left = CHILD_AT(exp, 0);
+	AlgebraicExpression *right = CHILD_AT(exp, 1);
+
+	GrB_Descriptor_new(&desc);  // Descriptor used for transposing operands.
+
+	if(left->type == AL_OPERATION) {
+		assert(left->operation.op = AL_EXP_TRANSPOSE);
+		GrB_Descriptor_set(desc, GrB_INP0, GrB_TRAN);
+		left = CHILD_AT(left, 0);
+	}
+	A = left->operand.matrix;
+
+	if(right->type == AL_OPERATION) {
+		assert(right->operation.op = AL_EXP_TRANSPOSE);
+		GrB_Descriptor_set(desc, GrB_INP1, GrB_TRAN);
+		right = CHILD_AT(right, 0);
+	}
+	B = right->operand.matrix;
+
+	if(B == IDENTITY_MATRIX) {
+		// B is the identity matrix, Perform A * I.
+		info = GrB_Matrix_apply(res, GrB_NULL, GrB_NULL, GrB_IDENTITY_BOOL, A, desc);
+		if(info != GrB_SUCCESS) {
+			// If the multiplication failed, print error info to stderr and exit.
+			fprintf(stderr, "Encountered an error in matrix multiplication:\n%s\n", GrB_error());
+			assert(false);
+		}
+	} else {
+		// Perform multiplication.
+		info = GrB_mxm(res, GrB_NULL, GrB_NULL, GxB_LOR_LAND_BOOL, A, B, desc);
+		if(info != GrB_SUCCESS) {
+			// If the multiplication failed, print error info to stderr and exit.
+			fprintf(stderr, "Encountered an error in matrix multiplication:\n%s\n", GrB_error());
+			assert(false);
+		}
+	}
+
+	GrB_Matrix_nvals(&nvals, res);
+
+	// Reset descriptor.
+	GrB_Descriptor_set(desc, GrB_INP0, GxB_DEFAULT);
+	GrB_Descriptor_set(desc, GrB_INP1, GxB_DEFAULT);
+
+	uint child_count = AlgebraicExpression_ChildCount(exp);
+	for(uint i = 2; i < child_count; i++) {
```

Comment:
Couldn't we start the loop at 1 and avoid the redundant logic for multiplying `right`?

Reply:
For similar reasons to `_Eval_Add` the first multiplication is done outside of the loop, to guarantee `res` is set, this allows us to skip multiplying by `identity` within the loop otherwise we'll need to keep track within the loop body rather or not `res` was set.
For the Identity to show up is not a common case, we could start the loop at 1 and always perform multiplication/apply. although I like the quick skip over `res * I` inside of the loop.